benchmarks/graphics_pipeline/main.cpp

```diff
@@ -701,49 +454,296 @@ void ProjApp::Setup()
+
+    bytecode = LoadShader("benchmarks/shaders", "Benchmark_RandomNoise.ps");
+    PPX_ASSERT_MSG(!bytecode.empty(), "PS shader bytecode load failed");
+    shaderCreateInfo = {static_cast<uint32_t>(bytecode.size()), bytecode.data()};
+    PPX_CHECKED_CALL(GetDevice()->CreateShaderModule(&shaderCreateInfo, &mPSNoise));
+
+    bytecode = LoadShader("benchmarks/shaders", "Benchmark_SolidColor.vs");
+    PPX_ASSERT_MSG(!bytecode.empty(), "VS shader bytecode load failed");
+    shaderCreateInfo = {static_cast<uint32_t>(bytecode.size()), bytecode.data()};
+    PPX_CHECKED_CALL(GetDevice()->CreateShaderModule(&shaderCreateInfo, &mVSSolidColor));
+
+    bytecode = LoadShader("benchmarks/shaders", "Benchmark_SolidColor.ps");
+    PPX_ASSERT_MSG(!bytecode.empty(), "PS shader bytecode load failed");
+    shaderCreateInfo = {static_cast<uint32_t>(bytecode.size()), bytecode.data()};
+    PPX_CHECKED_CALL(GetDevice()->CreateShaderModule(&shaderCreateInfo, &mPSSolidColor));
+}
 
-    // Graphics pipelines indexer
-    {
-        mGraphicsPipelinesIndexer.AddDimension(kAvailableVsShaders.size());
-        mGraphicsPipelinesIndexer.AddDimension(kAvailablePsShaders.size());
-        mGraphicsPipelinesIndexer.AddDimension(kAvailableVbFormats.size());
-        mGraphicsPipelinesIndexer.AddDimension(kAvailableVertexAttrLayouts.size());
+void ProjApp::SetupSkyboxMeshes()
+{
+    TriMesh  mesh = TriMesh::CreateCube(float3(1, 1, 1), TriMeshOptions().TexCoords());
+    Geometry geo;
+    PPX_CHECKED_CALL(Geometry::Create(GeometryOptions::InterleavedU16().AddTexCoord(), mesh, &geo));
+    PPX_CHECKED_CALL(grfx_util::CreateMeshFromGeometry(GetGraphicsQueue(), &geo, &mSkyBox.mesh));
+}
+
+void ProjApp::SetupSphereMeshes()
+{
+    // 3D grid
+    Grid grid;
+    grid.xSize = static_cast<uint32_t>(std::cbrt(kMaxSphereInstanceCount));
+    grid.ySize = grid.xSize;
+    grid.zSize = static_cast<uint32_t>(std::ceil(kMaxSphereInstanceCount / static_cast<float>(grid.xSize * grid.ySize)));
+    grid.step  = 10.0f;
+
+    // Get sphere indices
+    std::vector<uint32_t> sphereIndices(kMaxSphereInstanceCount);
+    std::iota(sphereIndices.begin(), sphereIndices.end(), 0);
+    // Shuffle using the `mersenne_twister` deterministic random number generator to obtain
+    // the same sphere indices for a given `kMaxSphereInstanceCount`.
+    Shuffle(sphereIndices.begin(), sphereIndices.end(), std::mt19937(kSeed));
+
+    // LODs for spheres
+    mSphereLODs.push_back(LOD{/* longitudeSegments = */ 50, /* latitudeSegments = */ 50, kAvailableLODs[0]});
+    mSphereLODs.push_back(LOD{/* longitudeSegments = */ 20, /* latitudeSegments = */ 20, kAvailableLODs[1]});
+    mSphereLODs.push_back(LOD{/* longitudeSegments = */ 10, /* latitudeSegments = */ 10, kAvailableLODs[2]});
+    PPX_ASSERT_MSG(mSphereLODs.size() == kAvailableLODs.size(), "LODs for spheres must be the same as the available LODs");
+
+    // Create the meshes
+    uint32_t meshIndex = 0;
+    for (LOD lod : mSphereLODs) {
```

Comment:
This is still way too much for a single function to do. We have to scroll 150+ lines to see from the end of this for loop to see how it started. The code will be read a lot more times than it is written, so would we be able to simplify organize this a bit?
- Put it into a separate file for geometry setup: for ex mesh_setup.h, mesh_setup.cpp
Add code in a new file where it only exposes a single function which returns the overall objects you need in the Application (which I think are sphere meshes):
```
static const GeometryOptions kLowPrecisionPositionPlanarOptions = GeometryOptions::PositionPlanarU32(grfx::FORMAT_R16G16B16_FLOAT)
                                                                .AddTexCoord(grfx::FORMAT_R16G16_FLOAT)
                                                                .AddNormal(grfx::FORMAT_R8G8B8A8_SNORM)
                                                                .AddTangent(grfx::FORMAT_R8G8B8A8_SNORM);

CreateSpheresForLod() {
	CreateSingleSpheresGeometries();
	PopulateSingleSpheresVertexData();
	CopySingleSphereToMultiSphere();
	UpdateSphereGeometryInPlace();

	for all meshes to create:
		grfx_util::CreateMeshFromGeometry(mesh);
}

CreateSphereMeshes() {
	// setup lods
	for (LOD lod : mSphereLods) {
		CreateSpheresForLod();
	}
}
```

or maybe even add
```
enum class PrecisionType {
    PRECISION_TYPE_LOW_PRECISION = 0;
};

enum class VertexLayoutType {
    VERTEX_LAYOUT_TYPE_INTERLEAVED = 0;
    VERTEX_LAYOUT_TYPE_PLANAR = 1;
};

GeometryOptions GetGeometryOptions(PrecisionType precisionType, VertexLayoutType vertexLayoutType) {
    
}
```

Reply:
I tried to refactor the sphere mesh code into a separate class, PTAL and let me know if this works